hpe3parclient/client.py

```diff
@@ -508,46 +508,66 @@ def createVolume(self, name, cpgName, sizeMiB, optional=None):
+                            "[%s], Valid values are[True, False]" % key
+                        raise exceptions.HTTPBadRequest(ex_desc)
+
+                combination = ['tdvv', 'compression']
+                len_diff = len(set(combination) - set(optional.keys()))
+                if len_diff == 1:
+                    raise exceptions.HTTPBadRequest(
+                        "Invalid input:For Deco volumes, "
+                        "both 'tdvv' and 'compression' must be specified")
+                if optional.get('tdvv') is True \
                         and optional.get('compression') is True:
                     optional['reduce'] = True
 
-                if not optional.get('tpvv') \
-                        and not optional.get('compression'):
-                    optional['tpvv'] = True
+                if optional.get('tdvv') is False \
+                        and optional.get('compression') is False:
+                    optional['reduce'] = False
 
-                if 'compression' in optional:
-                    if optional.get('compression') is not None:
-                        if optional.get('compression') is True:
-                            optional['reduce'] = True
-                        elif optional.get('compression') is False:
-                            optional['reduce'] = False
-                        else:
-                            # raising exception for junk compression input
-                            ex_desc = "39 - invalid input: wrong type for value \
-                                       - compression"
-                            raise exceptions.HTTPBadRequest(ex_desc)
+                if optional.get('tdvv') is False \
+                        and optional.get('compression') is True:
+                    raise exceptions.HTTPBadRequest(
+                        "Invalid input:For Deco volumes, "
+                        "both 'tdvv' and 'compression' must be specified")
 
-                    optional.pop('compression')
+                if optional.get('tdvv') is True \
+                        and optional.get('compression') is False:
+                    raise exceptions.HTTPBadRequest(
+                        "Invalid input:For Deco volumes, "
+                        "both 'tdvv' and 'compression' must be specified")
 
-                if optional.get('tpvv') is True \
-                        and optional.get('reduce') is True:
-                    optional.pop('tpvv')
+                if 'compression' in optional:
+                    optional.pop('compression')
+                if 'tdvv' in optional:
+                    optional.pop('tdvv')
             info = self._mergeDict(info, optional)
         logger.debug("Parameters passed for create volume %s" % info)
-
-        response, body = self.http.post('/volumes', body=info)
-        return body
+        try:
+            response, body = self.http.post('/volumes', body=info)
```

Comment:
@bhagyashree-sarawate @imran-ansari what happens when user tries to create volume by passing provisioning as full. How will your code behave along with Imran's changes #699 for plugin v2.

I do see we we pass full as provisioning then rest call will be done with tpvv as False and reduce as False.

Can you confirm the behavior from preimera for the same. Do we get the proper error or what kind of volume will get created.

Reply:
@nilangekarss If we pass provisioning as full, we get message as Invalid input received: invalid input: Either tpvv must be True OR tdvv and compression must be True. Both cannot be False.

Reply:
@nilangekarss 
from plugin side for primera
[root@cssosecostor-b10 ~]# date;docker volume create -d hpe --name bh_default1 -o provisioning=full
Fri Jul 26 12:57:26 IST 2019
Error response from daemon: create bh_default1: Invalid input received: For thin volumes, 'provisioning' must be set as 'thin'. And for deco volumes, 'provisioning' must be set as 'dedup' along with 'compression' set as true. If any of these conditions for a given type of volume is not met, volume creation will fail